File: cs/Loreline/Node.cs
using System;
using Runtime = Loreline.Runtime;
using Internal = Loreline.Internal;
using System.Collections.Generic;

namespace Loreline
{
    public struct NodeId
    {
        public readonly long Value;

        public NodeId(long value)
        {
            Value = value;
        }

        public static implicit operator NodeId(long value)
        {
            return new NodeId(value);
        }

        public static explicit operator long(NodeId id)
        {
            return id.Value;
        }

        public static explicit operator string(NodeId id)
        {
            return id.ToString();
        }

        public override string ToString()
        {
            return Runtime._Node.NodeId_Impl_.toString(Value);
        }
    }

    /// <summary>
    /// Represents a node in a Loreline AST.
    /// </summary>
    public class Node
    {
        /// <summary>
        /// The underlying runtime node instance.
        /// </summary>
        public readonly Runtime.Node RuntimeNode;

        /// <summary>
        /// The type of the node as string
        /// </summary>
        public readonly string Type;

        /// <summary>
        /// The id of this node (should be unique within a single script hierarchy)
        /// </summary>
        public readonly NodeId Id;

        /// <summary>
        /// The line number in the source code where this node appears (1-based).
        /// </summary>
        public readonly int Line;

        /// <summary>
        /// The column number in the source code where this node appears (1-based).
        /// </summary>
        public readonly int Column;

        /// <summary>
        /// The absolute character offset from the start of the source code.
        /// </summary>
        public readonly int Offset;

        /// <summary>
        /// The length of the source text span this node represents.
        /// A value of 0 indicates a point position rather than a span.
        /// </summary>
        public readonly int Length;

        public Node(Runtime.Node runtimeNode)
        {
            this.RuntimeNode = runtimeNode;
            this.Type = runtimeNode.type();
            this.Id = runtimeNode.id;
            var pos = runtimeNode.pos;
            this.Line = pos.line;
            this.Column = pos.column;
            this.Offset = pos.offset;
            this.Length = pos.length;
        }

        /// <summary>
        /// Converts the node to a JSON representation.
        /// This can be used for debugging or serialization purposes.
        /// </summary>
        /// <param name="pretty">Whether to format the JSON with indentation and line breaks</param>
        /// <returns>A JSON string representation of the node</returns>
        public string ToJson(bool pretty = false)
        {
            return Runtime.Json.stringify(RuntimeNode.toJson(), pretty);
        }
    }
}

File: jvm/src/loreline/Node.java
package loreline;

/**
 * Base class for all Loreline AST nodes.
 * Provides access to the node type, unique ID, and JSON export.
 */
public class Node {

    /**
     * The underlying runtime node instance.
     */
    public final loreline.runtime.Node runtimeNode;

    /**
     * The type of the node as a string (e.g. "Script", "Beat", "Text", "Dialogue").
     */
    public final String type;

    /**
     * The unique ID of this node within a single script hierarchy, as a raw Int64 value.
     * Use {@link #nodeIdToString()} for a human-readable representation.
     */
    public final long id;

    /**
     * The line number in the source code where this node appears (1-based).
     */
    public final int line;

    /**
     * The column number in the source code where this node appears (1-based).
     */
    public final int column;

    /**
     * The absolute character offset from the start of the source code.
     */
    public final int offset;

    /**
     * The length of the source text span this node represents.
     * A value of 0 indicates a point position rather than a span.
     */
    public final int length;

    Node(loreline.runtime.Node runtimeNode) {
        this.runtimeNode = runtimeNode;
        this.type = runtimeNode.type();
        this.id = runtimeNode.id.thisVal;
        loreline.runtime.Position pos = runtimeNode.pos;
        this.line = pos.line;
        this.column = pos.column;
        this.offset = pos.offset;
        this.length = pos.length;
    }

    /**
     * Returns the human-readable node ID string (e.g. "1.0.0.0").
     *
     * @return the node ID as a dotted string
     */
    public String nodeIdToString() {
        return runtimeNode.id.toString();
    }

    /**
     * Converts the node to a JSON string representation.
     * This can be used for debugging or serialization purposes.
     *
     * @param pretty whether to format the JSON with indentation and line breaks
     * @return a JSON string representation of the node
     */
    public String toJson(boolean pretty) {
        return loreline.runtime.Json.stringify(runtimeNode.toJson(), pretty);
    }

    /**
     * Converts the node to a compact JSON string representation.
     *
     * @return a JSON string representation of the node
     */
    public String toJson() {
        return toJson(false);
    }

    /**
     * Reconstructs a Node from a JSON string.
     *
     * @param json a JSON string (as returned by {@link #toJson()})
     * @return the reconstructed Node
     */
    public static Node fromJson(String json) {
        Object parsed = loreline.runtime.Json.parse(json);
        loreline.runtime.Node runtimeNode = loreline.runtime.Node.fromJson(parsed);
        return new Node(runtimeNode);
    }
}
